### PR TITLE
Sort primitives by size

### PR DESCRIPTION
**Pull request by Verkhovskaya**

### CODE REVIEW

File: src/geometry/PrimitiveGroup.ts [hunk 1]
```diff
@@ -102,20 +119,30 @@ export default abstract class PrimitiveGroup extends GeometryGroup {
...
-      this.treeIndexMap[value].push(index);
-    }
-
     this.treeIndex[index] = value;
   }
 
+  buildTreeIndexMap() {
+    this.treeIndexMap = {};
+    this.treeIndex.forEach((treeIndex, index) => {
+      if (index < this.data.count) {
```
Comment (anderscognite) [inline src/geometry/PrimitiveGroup.ts]: Can it ever happen that `index` is larger?

File: src/geometry/PrimitiveGroup.ts [hunk 2]
```diff
@@ -169,4 +196,19 @@ export default abstract class PrimitiveGroup extends GeometryGroup {
       itemSize: 3
     });
   }
+
+  sort() {
+    const newIndexes = this.data.sort();
```
Comment (anderscognite) [inline src/geometry/PrimitiveGroup.ts]:
```suggestion
    const newIndices = this.data.sort();
```

File: src/geometry/PrimitiveGroupData.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  sort() {
+    const sizeAndIndex: [number, number][] = [];
+    (this.arrays.size as Float32Array).forEach((size, index) => {
+      if (index < this.count) {
+        sizeAndIndex.push([size, index]);
+      }
+    });
+    const sortedIndexes = sizeAndIndex.sort(([size1], [size2]) => size2 - size1).map(([, index]) => index);
```
Comment (anderscognite) [inline src/geometry/PrimitiveGroupData.ts]:
```suggestion
    const sortedIndices = sizeAndIndex.sort(([size1], [size2]) => size2 - size1).map(([, index]) => index);
```

Comment (anderscognite) [reply src/geometry/PrimitiveGroupData.ts]: And update the usages later on of course!

File: src/geometry/PrimitiveGroup.ts (hunk 2 above)
Comment (anderscognite) [reply src/geometry/PrimitiveGroup.ts]: And update the usages after!

File: src/geometry/PrimitiveGroup.ts [hunk 4]
```diff
@@ -169,4 +196,19 @@ export default abstract class PrimitiveGroup extends GeometryGroup {
...
+    this.capacity = this.data.count;
+
+    const newTreeIndexes = new Float32Array(this.capacity);
```
Comment (anderscognite) [inline src/geometry/PrimitiveGroup.ts]:
```suggestion
    const newTreeIndices = new Float32Array(this.capacity);
```

File: src/parsers/i3d/unpackGeometry/main.ts [hunk 5]
```diff
@@ -151,10 +135,13 @@ function findOrCreateDestinationGroup(
   if (destinationGroup !== undefined) {
     return destinationGroup;
   } else {
+    let primitivesPerSectorAndChildren = 0;
```
Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/main.ts]: If this is a count, use `count` or `num`, or `numberOf` in the variable name maybe?

File: src/parsers/i3d/unpackGeometry/main.ts [hunk 6]
```diff
@@ -35,61 +34,46 @@ export function unpackPrimitives(
...
-  RenderedPrimitiveNames.forEach(renderedPrimitive => {
-    primitivesPerSectorAndChildren[sector.path][renderedPrimitive] = 0;
-  });
-
-  compressedData[sector.path].primitives.forEach(fileGeometryData => {
-    renderedPrimitivesPerFilePrimitive[fileGeometryData.type].forEach(renderedPrimitiveInfo => {
-      primitivesPerSectorAndChildren[sector.path][renderedPrimitiveInfo.name] +=
-        renderedPrimitiveInfo.count * fileGeometryData.count;
+function countRenderedPrimitivesPerSector(rootSector: Sector, compressedData: PerSectorCompressedData) {
+  const primitivesPerSector: PrimitivesPerSector = {};
```
Comment (anderscognite) [inline src/parsers/i3d/unpackGeometry/main.ts]: If this is a count, use `count` or `num`, or `numberOf` in the variable name maybe?

File: src/parsers/protobuf/main.ts [hunk 7]
```diff
@@ -157,13 +157,18 @@ export default async function parseProtobuf(
...
+  }
+
   mergeInstancedMeshes(rootSector, sceneStats);
   for (const sector of rootSector.traverseSectors()) {
     sector.mergedMeshGroup.createTreeIndexMap();
     sector.instancedMeshGroup.createTreeIndexMap();
 
     sceneStats.numSectors++;
     sector.primitiveGroups.forEach(primitiveGroup => {
+      // @ts-ignore
```
Comment (anderscognite) [inline src/parsers/protobuf/main.ts]: Why did we need a `ts-ignore` here? Could we fix it instead? If not, at least add the exception to the ignore statement.

File: src/parsers/protobuf/parseCircles.ts [hunk 8]
```diff
@@ -102,6 +101,7 @@ function parseConeEccentricConeCylinder(geometry: any[], group: CircleGroup, fil
     // @ts-ignore
   } else if (geometry.type === 'cone') {
     const { radiusA = 0, radiusB = 0 } = primitiveInfo;
+    const size = radiusA + radiusB;
```
Comment (anderscognite) [inline src/parsers/protobuf/parseCircles.ts]:
```suggestion
    const size = Math.max(2 * radiusA, 2 * radiusB);
```

File: src/parsers/protobuf/parseCircles.ts [hunk 9]
```diff
@@ -113,6 +113,7 @@ function parseConeEccentricConeCylinder(geometry: any[], group: CircleGroup, fil
     // @ts-ignore
   } else if (geometry.type === 'eccentricCone') {
     const { radiusA, radiusB } = primitiveInfo;
+    const size = radiusA + radiusB;
```
Comment (anderscognite) [inline src/parsers/protobuf/parseCircles.ts]:
```suggestion
    const size = Math.max(2 * radiusA, 2 * radiusB);
```

**Review by anderscognite — CHANGES_REQUESTED** [review]
Good stuff! Some naming issues (`indexes` vs `indices`) and other small comments, but nice!

File: src/geometry/PrimitiveGroup.ts (hunk 1 above)
Comment (Verkhovskaya) [reply src/geometry/PrimitiveGroup.ts]: Switched to a for loop

File: src/geometry/PrimitiveGroup.ts (hunk 2 above)
Comment (Verkhovskaya) [reply src/geometry/PrimitiveGroup.ts]: Fixed

File: src/geometry/PrimitiveGroup.ts (hunk 4 above)
Comment (Verkhovskaya) [reply src/geometry/PrimitiveGroup.ts]: Fixed

File: src/parsers/i3d/unpackGeometry/main.ts (hunk 5 above)
Comment (Verkhovskaya) [reply src/parsers/i3d/unpackGeometry/main.ts]: Fixed

File: src/parsers/protobuf/main.ts (hunk 7 above)
Comment (Verkhovskaya) [reply src/parsers/protobuf/main.ts]: Oh that's actually unnecessary. Removed.

File: src/geometry/PrimitiveGroup.ts (hunk 1 above)
Comment (anderscognite) [reply src/geometry/PrimitiveGroup.ts]: What do you mean? The code is identical, and I still don't understand how `index` could be larger than `this.data.count`?

File: src/geometry/PrimitiveGroup.ts (hunk 2 above)
Comment (anderscognite) [reply src/geometry/PrimitiveGroup.ts]: Still says indexes.

**Review by anderscognite — APPROVED** [review]
LGTM!

**Pull request merged**